l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+% Currently only a small subset of the functions provided by the \pkg{bitset}
+% package are implemented here, mainly the functions needed to use bitsets in
+% PDF dictionaries.
+%
+% The bitset is stored as a string and so the vector size is theoretically
+% unlimited, only restricted by \TeX-memory, but the export function
+% \cs{bitset_to_arabic:N} uses a function from the \texttt{int} module and
+% so the length is practically restricted to values smaller than 31 if this
+% output is needed.
+%
+% \section{Creating bitsets}
+% \begin{function}[added = 2020-12-13]
+%   {\bitset_new:N,\bitset_new:c}
+%   \begin{syntax}
+%     \cs{bitset_new:N}  \meta{bitset var} \\
+%   \end{syntax}
+% Creates a new \meta{bitset var} or raises an error if the name is already taken.
+% The declaration is global. The \meta{bitset var} is initially empty.
+%
+%  Bitsets are implemented as string variables consisting of
+%  \texttt{1}'s and \texttt{0}'s.
+%  The rightmost number is the index position 0, so
+%  the string variable can be viewed directly as the binary number.
+%
+% \end{function}
+%
+% \begin{function}[EXP, pTF,added = 2020-12-14]
+%   { \bitset_if_exist:N, \bitset_if_exist:c  }
+%   \begin{syntax}
+%     \cs{bitset_if_exist_p:N} \meta{bitset var}
+%     \cs{bitset_if_exist:NTF} \meta{bitset var} \Arg{true code} \Arg{false code}%
+%   \end{syntax}
+%   Tests whether the \meta{bitset var} exist.
+% \end{function}
+
+%
+% \section{Setting and clearing bits}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_set:Nn, \bitset_set:cn, \bitset_gset:Nn, \bitset_gset:cn  }
+%   \begin{syntax}
+%     \cs{bitset_set:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gset:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and sets the bit of the
+% index position it represents to 1. Index position are zero based and set from
+% right to left (so one can view the bitset as a binary number).
+% If needed the length of the bit vector is enlarged.
+% If the \meta{integer expression} evaluates to a negative number, nothing happens.
+% \end{function}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_clear:Nn, \bitset_clear:cn, \bitset_gclear:Nn, \bitset_clear:cn }
+%   \begin{syntax}
+%     \cs{bitset_clear:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gclear:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and clears the bit of the
+% index position it represents (sets it to 0).
+% The index is zero based. If the \meta{integer expression} evaluates to a
+% negative number or is larger than the current length of the bit vector
+% nothing happens. If the leading (left most) bit is cleared,
+% zeros are not trimmed but stay in the bit vector.
+% \end{function}
+%
+% \section{Using bitsets}
+%
+% \begin{function}[EXP,added = 2020-12-14]
+%   { \bitset_index:Nn, \bitset_index:cn }
```

Comment:
wouldn't that be confusing as the index is zero based?

Reply:
Ah, there's that... Okay, so `index` might be better

Though being zero-based is confusing by itself.  Does it have a strong reason for being so? (other than bits being _always_ zero-based everywhere :)

Reply:
> Though being zero-based is confusing by itself. Does it have a strong reason for being so? (other than bits being _always_ zero-based everywhere :)

While I never seen one-based bitsets in another context, the PDF specification uses them, so using zero-based indexing here means that in our primary usecase everything has to be shifted.

Reply:
oh yes, zero based is confusing. But as it it represents a (binary) number and the first (right) position is  for 2^0 there is imho not much one can do.

Reply:
@zauguin  oh my god, I have completly pushed that out of my mind. I'm so used to the numbering in hyperref and the bitset package, that I could have sworn the numering in the pdf reference starts with zero too. But I guess we still can't make it 1-based, or can we?

Reply:
I think we can go with indexing from 1. It's what we do for everything else, and it's not like people are (largely) going to do `2^(\bitset_index:n {position} )`. If we go with a zero index, we are almost certain to get confused ourselves as at least the 'always 1' plan is self-consistent.

On `index` vs `item`, I thought the point is that unlike say a `prop`, you don't have 'items' here as each entry is a flag. So `index` seemed closer to the idea we are dealing with. I think we need _some_ word there, we can't just have `\bitset:Nn`.

Reply:
if there is not a huge disadvantage in code I would go with indexing from 1 because it is consistent  with everything else we do. as to "item" or "index" as we support named positions I think item is again more consistent.

Reply:
> While I never seen one-based bitsets in another context, the PDF specification uses them, so using zero-based indexing here means that in our primary usecase everything has to be shifted.

one remark on that: we do have bithandling in LaTeX already (for the float specifiers), would be interesting to make sure that - at least in theory - all that is needed there could be properly expressed in that bitset code even if we do not change to it at this stage.

Reply:
the index is now 1-based, so it remains the question if the command should be `_index` or `item`  (this also affects the `named_index`  command).